File: Source/Extensions/Tessa.Extensions.Default.Client/UI/KrProcess/KrHideApprovalStagePermissionsDisclaimer.cs
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;

namespace Tessa.Extensions.Default.Client.UI.KrProcess
{
    public sealed class KrHideApprovalStagePermissionsDisclaimer :
        CardUIExtension
    {
        #region Private Methods

        private static void OnGridRowInvoked(object sender, GridRowEventArgs e)
        {
            if (e.Action != GridRowAction.Inserted && e.Action != GridRowAction.Opening)
            {
                return;
            }

            if (!e.RowModel.Controls.TryGet("DisclaimerControl", out IControlViewModel label))
            {
                return;
            }

            bool readOnly = e.RowModel.Card.Permissions.Resolver
                .GetRowPermissions(KrConstants.KrStages.Virtual, e.Row.RowID)
                .Has(CardPermissionFlags.ProhibitModify);

            if (readOnly)
            {
                string fieldName = StageTypeSettingsNaming.PlainColumnName("KrApprovalSettingsVirtual", "IsParallel");
                if (!e.Row.Get<bool>(fieldName))
                {
                    label.ControlVisibility = System.Windows.Visibility.Collapsed;
                    label.Block.RearrangeSelf();
                }
            }
            else if (e.RowModel.Controls.TryGet("IsParallelFlag", out IControlViewModel isParallelControl))
            {
                isParallelControl.PropertyChanged += (sender2, e2) =>
                {
                    if (e2.PropertyName == "IsChecked")
                    {
                        label.ControlVisibility =
                            (label.ControlVisibility == System.Windows.Visibility.Collapsed)
                                ? System.Windows.Visibility.Visible
                                : System.Windows.Visibility.Collapsed;

                        label.Block.RearrangeSelf();
                    }
                };
            }
        }

        #endregion

        #region Base Overrides

        public override async Task Initialized(ICardUIExtensionContext context)
        {
            ICardModel model = context.Model;

            if (model.CardType.Flags.HasNot(CardTypeFlags.AllowTasks))
            {
                return;
            }

            if (!model.Forms.TryGet("ApprovalProcess", out IFormWithBlocksViewModel approvalTab))
            {
                return;
            }

            foreach (IBlockViewModel block in approvalTab.Blocks)
            {
                if (block.Name == "ApprovalStagesBlock")
                {
                    foreach (IControlViewModel control in block.Controls)
                    {
                        if (control is GridViewModel grid)
                        {
                            grid.RowInvoked += OnGridRowInvoked;
                        }
                    }
                }
            }
        }

        #endregion
    }
}

File: WebClient SDK/src/default/ui/krProcess/krHideApprovalStagePermissionsDisclaimer.ts
import { reaction } from 'mobx';
import { CardUIExtension, ICardUIExtensionContext } from 'tessa/ui/cards';
import { hasNotFlag, hasFlag, Visibility } from 'tessa/platform';
import { CardTypeFlags } from 'tessa/cards/types';
import { GridViewModel, GridRowEventArgs, GridRowAction, CheckBoxViewModel } from 'tessa/ui/cards/controls';
import { CardPermissionFlags } from 'tessa/cards';

export class KrHideApprovalStagePermissionsDisclaimer extends CardUIExtension {

  private _disposes: (Function | null)[] = [];

  public initialized(context: ICardUIExtensionContext) {
    const model = context.model;

    if (hasNotFlag(model.cardType.flags, CardTypeFlags.AllowTasks)) {
      return;
    }

    const approvalTab = model.forms.find(x => x.name === 'ApprovalProcess');

    if (!approvalTab) {
      return;
    }

    for (let block of approvalTab.blocks) {
      if (block.name === 'ApprovalStagesBlock') {
        for (let control of block.controls) {
          if (control instanceof GridViewModel) {
            control.rowInvoked.add(this.gridRowInvoked);
          }
        }
      }
    }
  }

  public finalized() {
    for (let dispose of this._disposes) {
      if (dispose) {
        dispose();
      }
    }
    this._disposes.length = 0;
  }

  private gridRowInvoked = (e: GridRowEventArgs) => {
    if (e.action !== GridRowAction.Inserted && e.action !== GridRowAction.Opening) {
      return;
    }

    const label = e.rowModel!.controls.get('DisclaimerControl');
    if (!label) {
      return;
    }

    const readOnly = hasFlag(e.rowModel!.card.permissions.resolver
      .getRowPermissions('KrStagesVirtual', e.row.rowId), CardPermissionFlags.ProhibitModify);

    if (readOnly) {
      const fieldName = 'KrApprovalSettingsVirtual__IsParallel';
      if (!e.row.get(fieldName)) {
        label.controlVisibility = Visibility.Collapsed;
      }
    } else {
      const isParallelControl = e.rowModel!.controls.get('IsParallelFlag');
      if (isParallelControl) {
        this._disposes.push(reaction(
          () => (isParallelControl as CheckBoxViewModel).isChecked,
          () => {
            label.controlVisibility = label.controlVisibility === Visibility.Collapsed
              ? Visibility.Visible
              : Visibility.Collapsed;
          }
        ));
      }
    }
  }

}